src/unix/event.rs

```diff
@@ -0,0 +1,346 @@
+    }
+}
+
+impl IntoRawFd for EventNotifier {
+    fn into_raw_fd(self) -> std::os::unix::prelude::RawFd {
+        self.fd.into_raw_fd()
+    }
+}
+
+/// EventReceiver
+/// This is a generic event consumer that can be used with eventfd or pipefd.
+/// It allows reading a value from the file descriptor to consume an event.
+///
+/// # Examples
+///
+/// ```
+/// use std::os::fd::FromRawFd;
+/// use std::os::unix::io::IntoRawFd;
+/// use vmm_sys_util::event::EventConsumer;
+/// let (reader, _) = std::io::pipe().expect("Failed to create pipe");
+/// let consumer = unsafe { EventConsumer::from_raw_fd(reader.into_raw_fd()) };
+/// ```
+#[derive(Debug)]
+pub struct EventConsumer {
+    fd: File,
+}
+
+impl EventConsumer {
+    /// Read a value from the EventConsumer.
+    pub fn consume(&self) -> result::Result<(), io::Error> {
+        let mut buf = [0u8; size_of::<u64>()];
+        (&self.fd).read_exact(buf.as_mut_slice()).map(|_| Ok(()))?
+    }
+
+    /// Clone this EventConsumer.
+    pub fn try_clone(&self) -> result::Result<EventConsumer, io::Error> {
+        Ok(EventConsumer {
+            fd: self.fd.try_clone()?,
+        })
+    }
+}
+
+impl AsRawFd for EventConsumer {
+    fn as_raw_fd(&self) -> std::os::unix::prelude::RawFd {
+        self.fd.as_raw_fd()
+    }
+}
+
+impl FromRawFd for EventConsumer {
+    unsafe fn from_raw_fd(fd: std::os::unix::prelude::RawFd) -> Self {
+        EventConsumer {
+            fd: File::from_raw_fd(fd),
+        }
+    }
+}
+
+impl IntoRawFd for EventConsumer {
+    fn into_raw_fd(self) -> std::os::unix::prelude::RawFd {
+        self.fd.into_raw_fd()
+    }
+}
+
+#[cfg(not(any(target_os = "linux", target_os = "android")))]
+fn fcntl_getfl_and_setfl(fd: i32, flag: i32) -> std::result::Result<(), io::Error> {
+    // SAFETY: Safe because we check the fd is valid and check the return value.
+    let flags = unsafe { libc::fcntl(fd, libc::F_GETFL) };
+    if flags < 0 {
+        return Err(io::Error::last_os_error());
+    }
+    // SAFETY: Safe because we check the fd is valid and check the return value.
+    let ret = unsafe { libc::fcntl(fd, libc::F_SETFL, flags | flag) };
+    if ret < 0 {
+        return Err(io::Error::last_os_error());
+    }
+    Ok(())
+}
+
+fn fcntl_getfd_and_setfd(fd: i32, flag: i32) -> std::result::Result<(), io::Error> {
+    // SAFETY: Safe because we check the fd is valid and check the return value.
+    let flags = unsafe { libc::fcntl(fd, libc::F_GETFD) };
+    if flags < 0 {
+        return Err(io::Error::last_os_error());
+    }
+    // SAFETY: Safe because we check the fd is valid and check the return value.
+    let ret = unsafe { libc::fcntl(fd, libc::F_SETFD, flags | flag) };
+    if ret < 0 {
+        return Err(io::Error::last_os_error());
+    }
+    Ok(())
+}
+
+/// Create a new EventNotifier and EventConsumer using a pipe.
+///
+/// # Arguments
+///
+/// * `flag` - Flags to set on the file descriptor, such as `EventFlag::NONBLOCK` or `EventFlag::CLOEXEC`.
```

Comment:
nit: I would suggest renaming `flag` in `flags`, since also in the description we use the plural.

Please check all the places. (`EventFlag` is fine, but the variable can contain multiple flags, so I'd use the plural for the variable name)